### PR TITLE
snis: init at 20211017

### PR DESCRIPTION
**Pull request by cyclic-pentane**
###### Motivation for this change
[Space Nerds in Space](https://github.com/smcameron/space-nerds-in-space/) is a multiplayer spaceflight game that's (for example) already packaged in the [AUR](https://aur.archlinux.org/packages/snis-git/). It doesn't have any versioning, so the current date was chosen as the version of the derivation.

The derivation builds the 3D assets from the provided template files with `make models`. The makefile is patched so it uses Nixpkgs's `libopus` instead of downloading it as a zip archive while compiling. Additionally, the script `snis_text_to_speech.sh` is patched with the store paths of the binaries used.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/games/snis/default.nix [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+  nativeBuildInputs = [ pkg-config openscad ];
+  buildInputs = [ coreutils portaudio libbsd libpng libvorbis SDL2 lua5_2 glew openssl picotts sox alsa-utils libopus.dev ];
+
+  buildPhase = ''
+    runHook preBuild
+    make
+    make models
+    runHook postBuild
+  '';
```
Comment (SuperSandro2000) [inline pkgs/games/snis/default.nix:52]:
```suggestion
  postBuild = ''
    make models -j$NIX_BUILD_CORES
  '';
```

File: pkgs/games/snis/default.nix [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    substituteInPlace snis_launcher --replace "PREFIX=." "PREFIX=$out"
+    substituteInPlace snis_text_to_speech.sh --replace "pico2wave" "${sox}/bin/pico2wave" \
+                                             --replace "espeak" "${espeak-classic}/bin/espeak" \
+                                             --replace "play" "${sox}/bin/play" \
+                                             --replace "aplay" "${alsa-utils}/bin/aplay" \
+                                             --replace "/bin/rm" "${coreutils}/bin/rm"
+  '';
+
+  nativeBuildInputs = [ pkg-config openscad ];
+  buildInputs = [ coreutils portaudio libbsd libpng libvorbis SDL2 lua5_2 glew openssl picotts sox alsa-utils libopus.dev ];
```
Comment (SuperSandro2000) [inline pkgs/games/snis/default.nix:45]:
```suggestion
  buildInputs = [ coreutils portaudio libbsd libpng libvorbis SDL2 lua5_2 glew openssl picotts sox alsa-utils libopus ];
```

File: pkgs/games/snis/default.nix [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+  postPatch = ''
+    substituteInPlace Makefile --replace "OPUSARCHIVE=libopus.a" "OPUSARCHIVE=" \
+                               --replace "-I./opus-1.3.1/include" "-I${libopus.dev}/include/opus"
+    substituteInPlace snis_launcher --replace "PREFIX=." "PREFIX=$out"
+    substituteInPlace snis_text_to_speech.sh --replace "pico2wave" "${sox}/bin/pico2wave" \
+                                             --replace "espeak" "${espeak-classic}/bin/espeak" \
+                                             --replace "play" "${sox}/bin/play" \
+                                             --replace "aplay" "${alsa-utils}/bin/aplay" \
+                                             --replace "/bin/rm" "${coreutils}/bin/rm"
```
Comment (SuperSandro2000) [inline pkgs/games/snis/default.nix:41]:
```suggestion
    substituteInPlace Makefile \
      --replace "OPUSARCHIVE=libopus.a" "OPUSARCHIVE=" \
      --replace "-I./opus-1.3.1/include" "-I${libopus.dev}/include/opus"
    substituteInPlace snis_launcher --replace "PREFIX=." "PREFIX=$out"
    substituteInPlace snis_text_to_speech.sh \
      --replace "pico2wave" "${sox}/bin/pico2wave" \
      --replace "espeak" "${espeak-classic}/bin/espeak" \
      --replace "play" "${sox}/bin/play" \
      --replace "aplay" "${alsa-utils}/bin/aplay" \
      --replace "/bin/rm" "${coreutils}/bin/rm"
```

File: pkgs/games/snis/default.nix [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+, alsa-utils
+, espeak-classic
+, sox
+, libopus
+, openscad
+}:
+
+stdenv.mkDerivation {
+  pname = "snis_launcher";
+  version = "2021-10-17";
```
Comment (SuperSandro2000) [inline pkgs/games/snis/default.nix:24]:
```suggestion
  version = "unstable-2021-10-17";
```

**Comment by alyaeanyx** [comment]
Thanks for the suggestions, I've integrated them and squashed the commits into one

**Comment by tomberek** [comment]
I don't seem to be able to get SNIS to run via the launcher.

**Comment by alyaeanyx** [comment]
@tomberek What I did to test was:
`cd` into a directory in which the game files will be written
`nix shell github:alyaeanyx/nixpkgs/snis#snis`
`snis_launcher`
In the terminal, select Option 1, then 2, 3, and 4 (always taking the defaults). Then, a GUI window showing a 3D spaceship should pop up.

Would you mind sharing error logs/etc.?

**Comment by tomberek** [comment]
I am on wayland/sway so it's possible something is wonky. Yup:
```
Decoding audio data.../nix/store/dc67w77m3wcyyriq9mrp62i8gmvlasc7-snis_launcher-unstable-2021-10-17/share/snis/replacement_assets.txt: Warning:  No such file or directory
Done.
Reading solarsystem specifications...solar system asset file /nix/store/dc67w77m3wcyyriq9mrp62i8gmvlasc7-snis_launcher-unstable-2021-10-17/share/snis/solarsystems/default/assets.txt:ignoring line 6:atmosphere_brightness: 0.5
solar system asset file /nix/store/dc67w77m3wcyyriq9mrp62i8gmvlasc7-snis_launcher-unstable-2021-10-17/share/snis/solarsystems/default/assets.txt:ignoring line 8:atmosphere_brightness: 0.5
solar system asset file /nix/store/dc67w77m3wcyyriq9mrp62i8gmvlasc7-snis_launcher-unstable-2021-10-17/share/snis/solarsystems/default/assets.txt:ignoring line 10:atmosphere_brightness: 0.5
solar system asset file /nix/store/dc67w77m3wcyyriq9mrp62i8gmvlasc7-snis_launcher-unstable-2021-10-17/share/snis/solarsystems/default/assets.txt:ignoring line 12:atmosphere_brightness: 0.5
solar system asset file /nix/store/dc67w77m3wcyyriq9mrp62i8gmvlasc7-snis_launcher-unstable-2021-10-17/share/snis/solarsystems/default/assets.txt:ignoring line 14:atmosphere_brightness: 1.0
solar system asset file /nix/store/dc67w77m3wcyyriq9mrp62i8gmvlasc7-snis_launcher-unstable-2021-10-17/share/snis/solarsystems/default/assets.txt:ignoring line 16:atmosphere_brightness: 1.0
/nix/store/dc67w77m3wcyyriq9mrp62i8gmvlasc7-snis_launcher-unstable-2021-10-17/share/snis/solarsystems/default/assets.txt: expected 6 atmosphere brightnesses, but only found 0.
/nix/store/dc67w77m3wcyyriq9mrp62i8gmvlasc7-snis_launcher-unstable-2021-10-17/share/snis/solarsystems/default/assets.txt: planet texture ../../textures/planet-texture4- is type earthlike but has no normalmap
/nix/store/dc67w77m3wcyyriq9mrp62i8gmvlasc7-snis_launcher-unstable-2021-10-17/share/snis/solarsystems/default/assets.txt: planet texture ../../textures/planet-texture4- is type earthlike but has no normalmap
snis_client: solarsystem asset spec /nix/store/dc67w77m3wcyyriq9mrp62i8gmvlasc7-snis_launcher-unstable-2021-10-17/share/snis/solarsystems/default/assets.txt: 7 errors, 2 warnings.
done
Reading starbase model specifications...done
Could not create window: Could not get EGL display
```
This shouldn't be a blocker because it's probably due to my setup (though I'd love to figure it out, i get a similar error with Alacritty).

**Comment by alyaeanyx** [comment]
I've done a bit of superficial research, maybe https://wiki.archlinux.org/title/Wayland#SDL2 works for you? (SNIS uses SDL2). If it does, we could add a wrapper to the derivation that automatically adds that env var for all Wayland users

**Comment by tomberek** [comment]
Looks like the trick is to lie to it:
```
SDL_VIDEODRIVER=x11 snis_launcher
```
That worked for me, though I have no idea how to play and ran out of fuel. Thanks for packaging!

**Comment by alyaeanyx** [comment]
Thanks for your help with this issue! I've added a wrapper for the relevant binary.

**Pull request merged**